### ISSUE TITLE
[api-extractor] Any plans about extracting decorator annotations?

### ISSUE THREAD

**Issue by giniedp** [microsoft/rushstack#598]
The api-extractor is great. Unfortunately i am missing the ability to extract decorator annotations. When generating the docs i would be able to indicate which decorators have been used on a particular class or member. Are there any thoughts about this feature?

**Comment by pgonzal** [comment]
We originally went down this path, and we even created an [entire library](https://www.npmjs.com/package/@microsoft/decorators) for using decorators to express concepts such as \@sealed, \@virtual, \@override, etc.  

However, we ran into a problem that decorators don't appear in the \*.d.ts files.  The compiler doesn't consider them part of a type signature.  So we're now planning to convert these concepts into AEDoc comments instead of decorators.

When I asked the compiler group about it, they pointed out that decorators correspond to executable code.  For example, if you do this:

```typescript
class Blah {
  [memberId(123)]
  public doSomething(): void {
  }
}
```

...the "123" part could technically be something like this:

```typescript
class Blah {
  [memberId(Math.add(1,2,3,f(5)))]
  public doSomething(): void {
  }
}
```

They could emit a signature like this:

```typescript
class Blah {
  [memberId(number)]
  public doSomething(): void {
  }
}
```

...but then they would need to generate potentially very complex type expressions (e.g. if the parameter is a generic interface or something).

Also, decorators are still an experimental ECMA language feature.  

I'd be interested to hear your thoughts/scenario.

**Comment by giniedp** [comment]
true, they dont appear in d.ts. The API extractor does its job on d.ts files and not on the source .ts files, i didnt think about that.

My scenario is that i work on an Angular application where i implement lots of core components which are then used by other team members. Then i write up some examples and a little documentation for each component. The documentation consist of a summary block (which can be extracted from comments) and Angular specific properties (`@Input`, `@Output`, `@Component` decorators). Its tedious doing the docs by hand, so i hoped i could use the api-extractor.

As you can see, the `@sealed`, `@override`, `@virtual` keywords would not help me if they would be part of AEDoc.

**Comment by pgonzal** [comment]
> The API extractor does its job on d.ts files and not on the source .ts files, i didnt think about that.

Well, we could theoretically consider the source files, too.  I'm working on a feature that will map the error line numbers back to the \*.ts files, instead of showing line numbers of \*.d.ts files.  The reason for processing \*.d.ts files is technical:

- Some projects don't have any \*.ts files at all -- e.g. they generate TypeScript outputs from C# or some other system
- Some projects have complicated preprocessors that make it difficult for an intermediary stage like API Extractor to get the right compiler configuration for analyzing \*.ts files
- The \*.d.ts rollup generator is much simpler and executes faster when it processes small normalized \*.d.ts files instead of big \*.ts files with lots of implementation code

So that would be one approach.  

> i work on an Angular application where i implement lots of core components which are then used by other team members. Then i write up some examples and a little documentation for each component. The documentation consist of a summary block (which can be extracted from comments) and Angular specific properties (@Input, @Output, @Component decorators). Its tedious doing the docs by hand, so i hoped i could use the api-extractor.

The other approach would be to try again to persuade the compiler group to consider decorators as part of the type signature.  This is certainly the case for Angular.  From a broader perspective, how are consumers of your library supposed to understand the @Input/@Output/@Component contracts if they don't even appear in the \*.d.ts files?  They would have to go read your source code, assuming your library is open source.

The TypeScript compiler now has more elaborate plugins for transformations.  Perhaps you could make a plugin which copies the decorator expressions into code comments.

@RyanCavanaugh FYI

**Comment by giniedp** [comment]
The line number would be very useful for generating links to github (or whatever). If you could fish for Decorators, that would be nice but actually doesn't feel right.

As long as decorators are not part of d.ts spec i favor an option for ApiExtractor to select whether it scans the d.ts files or raw ts source files. How much of a "big deal" would that be? I mean, no mercy for fancy project setup and all that. Just as if the ts files were d.ts files.

I've written a simple extractor that fits my needs (for the moment). About 200 lines of code and is obviously not as powerful as AE. But gives me an impression of how fast the ts tools actually are. 200000LOC in 1500 files execute in less than 3 seconds. I wouldn't mind running AE on ts files.

**Comment by pgonzal** [comment]
> As long as decorators are not part of d.ts spec i favor an option for ApiExtractor to select whether it scans the d.ts files or raw ts source files. How much of a "big deal" would that be? I mean, no mercy for fancy project setup and all that. Just as if the ts files were d.ts files.

I'm still undecided on this issue.  It seems like a good idea, but processing *.d.ts files has certain advantages that seem to be locking us in to this approach.  A recent example:

Migrating a large production application to the latest compiler is typically a nontrivial amount of work, so it gets procrastinated and our projects often lag behind the latest release by 6 months or more. But we're now trying to keep API Extractor on the latest TypeScript version, which means we use it to analyze projects built using an older compiler.  There are a number of cases where TypeScript 3.x cannot compile source code that was written for TypeScript 2.x, but our approach mostly works in practice because most of the problematic constructs are normalized in the *.d.ts files.

**Comment by pgonzal** [comment]
@giniedp After further thought, we're not planning to support analysis of \*.ts files for now.  There are just too many advantages to working with the output \*.d.ts files.

So if we wanted to analyze decorators, some options:
1. The compiler would need to provide a way to include this information in the \*.d.ts files that it emits (which makes sense, as it's clearly part of your API contract); OR
2. API Extractor would need to use [declaration maps](https://github.com/Microsoft/web-build-tools/issues/894) to grab the extra metadata from the source files; OR
3. An easier (but less ideal) possibility would be to encode the same information in the doc comment (e.g. if the `@sealed` decorator is applied to a function, then the `@sealed` TSDoc tag should be included in the comment). Custom TSDoc tags could be defined for each decorator.

Let me know if you have any thoughts on this.

**Comment by octogonz** [comment]
Closing since there doesn't seem to be a straightforward way to implement this.  Feel free to reopen the issue if you can propose something.

**Comment by justinfagnani** [comment]
I maintain libraries that use decorators extensively to register components, create public observable properties, etc. It would be great to be able to use api-extractor's JSON output to collect this info for docs, but obviously I can't yet.

What are the problems of allowing entrypoints to be .ts files instead of .d.ts files? If an invocation opted into this, and it was slower, then that's really on the user, not api-extractor. My experience with tsc API so far shows there's not a lot of difference in working with the two formats.

**Comment by zelliott** [comment]
Given #894 is fixed, is it possible to go with #2 (using declaration maps to grab the decorators from the source)?

**Comment by octogonz** [comment]
This has a number of downsides. It might be slow, since API Extractor needs to read both .d.ts and .ts files. Also, the source maps are not particularly reliable for non-executable code (try loading them into [this tool](https://sokra.github.io/source-map-visualization/) to see what I mean).  It's also an optional feature -- API Extractor is expected to work without source maps.

>  as they are an important part of an Angular component's API.

I agree. We should solve it somehow.  However decorators are an important part of the API, shouldn't they appear in the emitted `.d.ts` file?  That is the contract that developers will normally see when they consume your library.  (It's not common practice for a shipping NPM package to include the `.ts` source code.)

If we want to pursue this, maybe a better approach would be to try go get the API information emitted in the .d.ts file. For example, maybe it could be a compiler plugin.  (?)

**Comment by justinfagnani** [comment]
@octogonz how would libraries that don't have their own compiler do this? We use decorators and plain tsc.

**Comment by octogonz** [comment]
I feel somewhat strongly that if decorators are part of your API contract, then they belong in the .d.ts file.  How else are API consumers supposed to know about it?

If there isn't already a GitHub issue proposing this featuire, someone should [create one](https://github.com/microsoft/TypeScript/issues/new/choose).  (And please CC me so I can follow along. :-) )

Without modify the compiler, there's probably other creative ways to get the information into the .d.ts file.  🤔 What if we proposed a TSDoc tag that would allow you to manually add these annotations?

For example:
```ts
class Point {
  private _x: number;
  constructor(x: number, y: number) {
    this._x = x;
  }

  /**
   * The x location of the point, measured in integer pixels.
   * @decorator `@configurable(false)`
   */
  @configurable(false)
  get x() {
    return this._x;
  }
}
```
It's a little awkward to duplicate the decorator like this.  But it has some advantages:
- Decorators are executable code that can be arbitrarily complex expressions (`@configurable('xyz', f(2*3,this))`), so maybe in some cases you'd want to write it more concisely for the docs (`@configurable('xyz', ...)`)
- Decorators are an experimental language feature, whose spec keeps getting rejected and reworked, so this would avoid any compatibility issues that might cause.

**Comment by octogonz** [comment]
Here's a TSDoc RFC: https://github.com/microsoft/tsdoc/issues/271

**Comment by octogonz** [comment]
Reopening this issue, since we're going to solve it using the `@decorator` TSDoc tag.

**Comment by octogonz** [comment]
Here's a PR: https://github.com/microsoft/rushstack/pull/2377

**Comment by justinfagnani** [comment]
Hmm... this doesn't seem to be a full solution for our use case. We don't want to require that our users add both a decorator and a `@decorator` JSDoc tag to their classes. The existence of the decorator itself should be enough information to confer what's needed for our tools.

**Comment by octogonz** [comment]
> If there isn't already a GitHub issue proposing this featuire, someone should [create one](https://github.com/microsoft/TypeScript/issues/new/choose). (And please CC me so I can follow along. :-) )

**Comment by octogonz** [comment]
This fix was released with API Extractor 7.12.1 / API Documenter 7.12.0
